cmd/checkpoint/main.go

```diff
@@ -0,0 +1,268 @@
+	return strings.Contains(pod.Name, "kube-apiserver") && pod.Namespace == api.NamespaceSystem
+}
+
+func isTempAPI(pod v1.Pod) bool {
+	return strings.Contains(pod.Name, "temp-apiserver") && pod.Namespace == api.NamespaceSystem
+}
+
+// cleanVolumes will sanitize the list of volumes and volume mounts
+// to remove the default service account token.
+func cleanVolumes(p *v1.Pod) {
+	volumes := make([]v1.Volume, 0, len(p.Spec.Volumes))
+	for _, v := range p.Spec.Volumes {
+		if !strings.Contains(v.Name, "default") {
+			volumes = append(volumes, v)
+		}
+	}
+	p.Spec.Volumes = volumes
+	for i := range p.Spec.Containers {
+		c := &p.Spec.Containers[i]
+		volumeMounts := make([]v1.VolumeMount, 0, len(c.VolumeMounts))
+		for _, vm := range c.VolumeMounts {
+			if !strings.Contains(vm.Name, "default") {
+				volumeMounts = append(volumeMounts, vm)
+			}
+		}
+		c.VolumeMounts = volumeMounts
+	}
+}
+
+func modifyInsecurePort(p *v1.Pod) {
+	for i := range p.Spec.Containers {
+		c := &p.Spec.Containers[i]
+		cmds := c.Command
+		for i, cmd := range cmds {
+			if strings.Contains(cmd, "insecure-port") {
+				cmds[i] = strings.Replace(cmd, "8080", "8081", 1)
+				break
+			}
+		}
+	}
+}
+
+// writeManifest will write the manifest to the ignore path.
+// It first writes the file to a temp file, and then atomically moves it into
+// the actual ignore path and correct file name.
+func writeManifest(manifest v1.Pod) {
+	m, err := json.Marshal(manifest)
+	if err != nil {
+		log.Fatal(err)
+	}
+	writeAndAtomicCopy(m, ignoreFile)
+}
+
+func parseAPIPodSpec(podList v1.PodList) v1.PodSpec {
+	var apiPod v1.Pod
+	for _, p := range podList.Items {
+		if strings.Contains(p.Name, string(kubeAPIServer)) {
+			apiPod = p
+			break
+		}
+	}
+	cleanVolumes(&apiPod)
+	modifyInsecurePort(&apiPod)
+	return apiPod.Spec
+}
+
+func newAPIClient() clientset.Interface {
+	kubeConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(&clientcmd.ClientConfigLoadingRules{
+		ExplicitPath: kubeconfigPath},
+		&clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+	client, err := clientset.NewForConfig(&restclient.Config{
+		Host: secureAPIAddr,
+		TLSClientConfig: restclient.TLSClientConfig{
+			CertData: kubeConfig.CertData,
+			KeyData:  kubeConfig.KeyData,
+			CAData:   kubeConfig.CAData,
+		},
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+	return client
+}
+
+func convertSecretsToVolumeMounts(client clientset.Interface, spec *v1.PodSpec) {
+	log.Println("converting secrets to volume mounts")
+	for i := range spec.Volumes {
+		v := &spec.Volumes[i]
+		if v.Name == "secrets" && v.Secret != nil {
+			v.HostPath = &v1.HostPathVolumeSource{
+				Path: tlsPath,
+			}
+			copySecretsToDisk(client, v.Secret.SecretName)
+			v.Secret = nil
+			break
+		}
+	}
+}
+
+func copySecretsToDisk(client clientset.Interface, name string) {
+	log.Println("copying secrets to disk")
+	if err := os.MkdirAll(tlsPath, 0755); err != nil {
```

Comment:
Can we try and follow the (sort of) existing secrets convention. Maybe:

`/var/run/secrets/coreos.com/checkpoints/<pod-name>/<secret-name>`

Eventually we could change <pod-name> to maybe the hash of the pod so we kept secrets in sync with pod checkpoints and could be GC'd etc.

Which brings up another side-point: What about cleaning up old secrets that no longer exist for the current checkpoint? This could be a follow-up issue (or just put TODO in code).